core/src/main/java/org/apache/iceberg/util/LocationUtil.java

```diff
@@ -33,4 +34,8 @@ public static String stripTrailingSlash(String path) {
     }
     return result;
   }
+
+  public static String posixNormalize(String path) {
+    return Paths.get(path).normalize().toString();
```

Comment:
This method is not suitable for Iceberg paths, since Java `Path` is for paths, but Iceberg paths are actually URIs. It mangles URIs and doesn't understand the scheme vs path distinction:
```
jshell> Paths.get("s3://foo/bar").toString()
$2 ==> "s3:/foo/bar"

jshell> Paths.get("s3://foo/bar/../..").normalize().toString()
$5 ==> "s3:"

jshell> Paths.get("s3://foo/bar/../../..").normalize().toString()
$6 ==> ""
```

`URI.normalize()` seems better, though it has interesting behavior where it preserves leading `..` path segments at the start of the path that would go past the root:

```
jshell> URI.create("s3://foo/abc/../../..").normalize()
$11 ==> s3://foo/../..
```

Reply:
Looks like using URI is still not the right way, because URI would fail with characters like space, which are supposed to be escaped with URL encoding like `%20`

I think what should be done is that, for path with scheme and optional authority, everything after that should use `Paths.get(path).normalize().toString()`.